### PR TITLE
Discovery - a fresh approach to preferences and typeIndexes

### PR DESCRIPTION
**Pull request by timbl**
discoveryLogic is a start from scratch ... with communities built  into it.
So you can do discovery on other people not just logged on user.

### CODE REVIEW

File: src/discovery/discoveryLogic.ts [hunk 1]
```diff
@@ -1,90 +1,247 @@
...
+        if (err.response.status === 404) {
+          debug.log('createIfNotExists doc does NOT exist, will create: ' + doc)
+          try {
+            store.fetcher.webOperation('PUT', doc.uri, {data: '', contentType: 'text/turtle'})
+          } catch (err) {
+            const msg = 'createIfNotExists doc FAILED: ' + doc + ': ' + err
+            debug.log(msg)
+            throw new Error(msg)
+          }
+          delete store.fetcher.requested[doc.uri] // delete cached 404 error
```
Comment (timea-solid) [inline src/discovery/discoveryLogic.ts:37]: I did not know about this call. I haven't seen it so far. 
Question: don't we have to load the newly created resource into the cache anymore?
I expect this mostly from the name of the function "load". Of course, it depends on what 'load' means -> load in the cache presumably. And it also depends if the creation automatically loads in cache beforehand. 
So what is the truth?

File: src/discovery/discoveryLogic.ts [hunk 2]
```diff
@@ -1,90 +1,247 @@
...
+
+}
+/* Follow link from this doc to another thing, or else make a new link
+**
+**  null return means new one created
+*/
+export async function followOrCreateLink(store: LiveStore, subject: NamedNode, predicate: NamedNode,
+     object: NamedNode, doc:NamedNode):Promise<NamedNode | null> {
+  const result = store.any(subject, predicate, null, doc)
+  if (result) return result as NamedNode
```
Comment (timea-solid) [inline src/discovery/discoveryLogic.ts:78]: Shouldn't we be interested to check if result === object? 
What if they are not equal?

File: src/discovery/discoveryLogic.ts [hunk 3]
```diff
@@ -1,90 +1,247 @@
...
-    const priv = privateTypeIndexes.length > 0 ? [ { label: 'priSo @@@@@vate', index: privateTypeIndexes[0] as NamedNode, agent: user } ] : []
-    return pub.concat(priv)
+     // console.log('@@ loadTypeIndexesFor privateTypeIndexes', privateTypeIndexes)
+
+    priv = privateTypeIndexes.length > 0 ? [ { label: 'private', index: privateTypeIndexes[0] as NamedNode, agent: user } ] : []
+  } else {
+    priv = []
+  }
+  const scopes =  publicScopes.concat(priv)
+  if (scopes.length === 0) return scopes
```
Comment (timea-solid) [inline src/discovery/discoveryLogic.ts:144]: If there is no `publicTypeIndex` we do not continue to private? Should it be a precondition?

File: src/discovery/discoveryLogic.ts [hunk 4]
```diff
@@ -1,90 +1,247 @@
...
 
 export async function loadAllTypeIndexes (store:LiveStore, user:NamedNode) {
   return (await loadTypeIndexesFor(store, user)).concat((await loadCommunityTypeIndexes(store, user)).flat())
 }
 
-/*
-export async function getAppInstances (store:LiveStore, klass: NamedNode) {
+// Utility: remove duplicates from Array of NamedNodes
 
+export function uniqueNodes (arr: NamedNode[]): NamedNode[] {
```
Comment (timea-solid) [inline src/discovery/discoveryLogic.ts:177]: This can be moved later to /util/UtilityLogic.ts

File: src/discovery/discoveryLogic.ts [hunk 5]
```diff
@@ -1,90 +1,247 @@
...
+  const uris = arr.map(x => x.uri)
+  const set = new Set(uris)
+  const uris2 = Array.from(set)
+  const arr2 = uris2.map(u => new NamedNode(u))
+  return arr2 // Array.from(new Set(arr.map(x => x.uri))).map(u => sym(u))
 }
-*/
+
+export async function getScopedAppsFrommIndex (store, scope, theClass: NamedNode) {
```
Comment (timea-solid) [inline src/discovery/discoveryLogic.ts:185]: please rename -> functions contains double `m` in getScopedAppsFro`mm`Index ?

File: src/discovery/discoveryLogic.ts [hunk 6]
```diff
@@ -1,90 +1,247 @@
...
+  const index = scope.index
+  const registrations = store.each(undefined, ns.solid('forClass'), theClass, index)
+  // console.log('    registrations', registrations )
+
+  // In practice it looks as though existing code does not put in the type explicitly so can't do this filter. Discuss.
+  // const relevant = registrations.filter(reg => store.holds(reg, ns.rdf('type'), ns.solid('TypeRegistration'), index))
+  // console.log('    relevant', relevant )
+  const relevant = registrations
+
+  const directInstances = registrations.map(reg => store.any(reg as NamedNode, ns.solid('instance'), null, index))
```
Comment (timea-solid) [inline src/discovery/discoveryLogic.ts:196]: So we expect there to be only one instance or container per type. Correct?

**Comment by timea-solid** [comment]
Much leaner and clearer code. 
I will ask in the team meeting too: what are community type indexes?

File: src/discovery/discoveryLogic.ts (hunk 6 above)
Comment (timbl) [reply src/discovery/discoveryLogic.ts:215]: Yup

**Comment by NoelDeMartin** [comment]
Hey, I know this has been merged already but we talked about doing a review in the SolidOS meeting, so I'm leaving my feedback here.

Overall I think it's quite easy to understand and readable, here's some things that come to mind:

- I'd avoid having commented code in the repository, it's ok to have it locally but it shouldn't be committed. Things like the commented `console.log` or try/catch hurt readability.
- It seems like type indexes are created empty and later populated. It'd be better to create the files directly with the type registrations to reduce network requests and possible conflicts.
- There is a `registerInstanceInTypeIndex` method but not `registerInstanceContainerInTypeIndex`? I guess it depends on the app, but in my apps so far I've only been using container registrations. I guess if this is intended for reuse, it should implement both.
- The `getScopedAppInstances` method name is a bit confusing. After looking at the implementation I think I understand what it does (getting instances registered in the type index). But I'm not sure what this has to do with an "App" (it made me think of the trusted apps registration, but I don't think it's related).
- I've noticed that this also creates a preferences file, but so far I haven't used that file in my apps at all, I've only been looking at the profile document. I also didn't know anything about community type indexes. I guess I should learn more about this, and it seems like part of it is already addressed [in the draft](https://github.com/solid/webid-profile/blob/main/notes/pre-final-draft.md) :).

Something I also mentioned in the call is that I probably wouldn't use this in my apps because it requires rdflib, which is a heavy dependency that I'm not using. So far I've implemented this just using fetch requests (and [N3.js](https://github.com/rdfjs/N3.js) to parse turtle).

In case you're interested to see how I've implemented it, you can find it in [@noeldemartin/solid-utils](https://github.com/NoelDeMartin/solid-utils/blob/main/src/helpers/interop.ts#L19), and here's how I'm using it in [Media Kraken](https://github.com/NoelDeMartin/media-kraken/blob/main/src/models/users/SolidUser.ts#L139). As I mentioned, there are some things I just learned by looking at this code so I'll probably change mine at some point.

**Pull request merged**